phoenix-core/src/main/java/org/apache/phoenix/query/ConnectionQueryServicesImpl.java

```diff
@@ -2405,16 +2413,26 @@ public Void call() throws Exception {
                             openConnection();
                             hConnectionEstablished = true;
                             boolean isDoNotUpgradePropSet = UpgradeUtil.isNoUpgradeSet(props);
+                            boolean doesSystemCatalogAlreadyExist = false;
```

Comment:
The flow of this code is very confusing. Let's do the compatibility check only once in ensureTableCreated by making the following changes:
- do not call ensureTableCreated from createTable if the table is a system table
- call ensureTableCreated instead here and remove this entire try block
- do all checks required in ensureTableCreated where we already determine if the table exists or not
- in ensureTableCreated, if SYSTEM.CATALOG doesn't exist and !isAutoUpgradeEnabled or isDoNotUpgradePropSet, then throw our standard UpgradeRequiredException immediately without creating system catalog metadata.
- the only call to checkClientServerCompatibility should be in ensureTableCreated (when isMetaTable is true)
- make sure to be defensive in the creation of the SYSTEM namespace and moving of SYSTEM tables as it's possible that multiple clients may be attempting to do that.

I think this improve the maintainability of this code (and fix this issue too).

Reply:
I'd also recommend committing PHOENIX-4668 and PHOENIX-4575 first as that'll simplify what you need to do here a bit.

Reply:
@JamesRTaylor thanks for the review. I've submitted a patch for PHOENIX-4668 and will go ahead with the changes you've mentioned for this patch. I'm not sure about the status of PHOENIX-4575 (@mujtabachohan was working on this before his vacation), so don't want to make any changes to it without understanding its scope.
Thanks.

Reply:
@JamesRTaylor A few follow-up questions from your comment:

- If we were to call _ensureTableCreated_ at this point inside the init method, don't have the values of other arguments such as the families, splits, etc. which we would need for the admin.createTable call. Instead I propose splitting the _ensureTableCreated_ method and having another private method which just checks for the existence of system catalog and does client-server compatibility checks. We can reuse this stub inside _ensureTableCreated_ and make sure we don't do this again when called via _createTable_.
-  If we throw an UpgradeRequiredException in case an upgrade is required, the user will not get the connection object right? This also prevents the user from being able to run "EXECUTE UPGRADE". Instead I guess we can call _setUpgradeRequired_, log an error and return the connection object. This will disallow the user from running anything except "EXECUTE UPGRADE" since we have this snippet in place:
`                        
if (conn.getQueryServices().isUpgradeRequired() && !conn.isRunningUpgrade()
    && stmt.getOperation() != Operation.UPGRADE) {
        throw new UpgradeRequiredException();
}
`

Reply:
@JamesRTaylor @twdsilva also, do we really want to avoid calling ensureTableCreated from createTable in case it is a system table? Apart from actually ensuring that the table is created or not and the client-server compatibility checks, this method also modifies the table according to a new descriptor (I don't have enough background as to why we do this here). Any advice guys?

Reply:
Good points. How about if we leave the ensureTableCreated call where it is, but move the logic in that try block of the init method into the ensureTableCreated method so it’s consolidated?

Reply:
@JamesRTaylor that could be a possible solution, but then we would always go through executeUpdate-> ... -> _MetaDataClient.createTable_  -> _MetaDataClient.createTableInternal_ -> _CQSI.creataTable_ before finally calling ensureTableCreated and a lot of unnecessary steps would be taken till that point. If we were to disallow the user to proceed in creating/upgrading SYSTEM.CATALOG inside ensureTableCreated, then this would lead to a lot of unnecessary intermediary steps. Instead, can we put all the necessary conditions inside that try block in a separate method which is always only called inside init? This method would also call _checkClientServerCompatibility_ and thus would fail fast in case an upgrade is required, avoiding other unnecessary steps like updating the cache, etc.